src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -241,6 +280,240 @@ void DiscoveryDataBase::AckedFunctor::operator () (
+    {
+        for (auto wit = tit->second.begin(); wit != tit->second.end(); ++wit)
+        {
+            if(wit->guidPrefix == participant_guid.guidPrefix)
+            {
+                tit->second.erase(wit);
+                --wit;
+            }
+        }
+    }
+
+    // Remove participant from others participants_[]::relevant_participants_builtin_ack_status
+    for (auto pit = participants_.begin(); pit != participants_.end(); ++pit)
+    {
+        pit->second.remove_participant(participant_guid.guidPrefix);
+    }
+
+    // Remove participant from others writers_[]::relevant_participants_builtin_ack_status
+    for (auto wit = writers_.begin(); wit != writers_.end(); ++wit)
+    {
+        wit->second.remove_participant(participant_guid.guidPrefix);
+    }
+
+    // Remove participant from others readers_[]::relevant_participants_builtin_ack_status
+    for (auto pit = readers_.begin(); pit != readers_.end(); ++pit)
+    {
+        pit->second.remove_participant(participant_guid.guidPrefix);
+    }
+
+    if(std::find(disposals_.begin(), disposals_.end(), ch) == disposals_.end())
+    {
+        disposals_.push_back(ch);
+    }
+}
+
+void DiscoveryDataBase::create_participant_from_change(
+        eprosima::fastrtps::rtps::CacheChange_t* ch)
+{
+    DiscoveryParticipantInfo part(ch);
+    participants_.insert(std::make_pair(guid_from_change(ch).guidPrefix, part));
+}
+
+void DiscoveryDataBase::create_writers_from_change(
+        eprosima::fastrtps::rtps::CacheChange_t* ch,
+        const eprosima::fastrtps::string_255& topic_name)
+{
+    const eprosima::fastrtps::rtps::GUID_t& writer_guid = guid_from_change(ch);
+
+    DiscoveryEndpointInfo tmp_writer(ch, topic_name);
+    std::pair<std::map<eprosima::fastrtps::rtps::GUID_t,DiscoveryEndpointInfo>::iterator,bool> ret;
+    ret = writers_.insert(std::make_pair(writer_guid, tmp_writer));
+
+    if (ret.second) {
+        auto readers_it = readers_by_topic_.find(topic_name);
+        if (readers_it != readers_by_topic_.end())
+        {
+            for (auto reader_it: readers_it->second)
+            {
+                writers_.find(writer_guid)->second.add_or_update_ack_participant(reader_it.guidPrefix);
+
+                auto rit = readers_.find(reader_it);
+                if (rit != readers_.end())
+                {
+                    rit->second.add_or_update_ack_participant(writer_guid.guidPrefix);
+                }
+
+                auto pit = participants_.find(reader_it.guidPrefix);
+                if (pit != participants_.end())
+                {
+                    pit->second.add_or_update_ack_participant(writer_guid.guidPrefix);
+                }
+            }
+        }
+
+        auto pit = participants_.find(writer_guid.guidPrefix);
+        if (pit != participants_.end())
+        {
+            pit->second.add_writer(writer_guid);
+        }
+
+        writers_by_topic_[topic_name].push_back(writer_guid);
```

Comment:
Won't this have problems with updates? I'd check whether is there or not before pushing_back. I'm thinking something like (not tested):
```suggestion
auto topic_it =  std::find(
            writers_by_topic_[topic_name].begin(),
            writers_by_topic_[topic_name].end(),
            topic_name);

        // New writer in the topic
        if (topic_it == writers_by_topic_[topic_name].end())
        {
            writers_by_topic_[topic_name].push_back(writer_guid);
        }
        // Update writer in the topic
        else
        {
            *topic_it = writer_guid;
        }
```